internal/pkg/addon/output.go

```diff
@@ -25,52 +29,127 @@ type Output struct {
-	//
-	// Instead, we want to retain the input logical ID for "Ref" so that we can check if an output refers
-	// to a particular CFN resource type.
-	tpl, err := goformation.ParseYAMLWithOptions([]byte(template), &intrinsics.ProcessorOptions{
-		IntrinsicHandlerOverrides: map[string]intrinsics.IntrinsicHandler{
-			// Given an output with "Value: !Ref AdditionalResourcesPolicy",
-			// this override evaluates to "Value: AdditionalResourcesPolicy".
-			"Ref": func(_ string, input interface{}, _ interface{}) interface{} {
-				if logicalID, ok := input.(string); ok {
-					return logicalID
-				}
-				return nil
-			},
-		},
-	})
+	type cfnTemplate struct {
+		Resources yaml.Node `yaml:"Resources"`
+		Outputs   yaml.Node `yaml:"Outputs"`
+	}
+	var tpl cfnTemplate
+	if err := yaml.Unmarshal([]byte(template), &tpl); err != nil {
+		return nil, fmt.Errorf("unmarshal addon cloudformation template: %w", err)
+	}
+
+	typeFor, err := parseTypeByLogicalID(&tpl.Resources)
+	if err != nil {
+		return nil, err
+	}
+
+	outputNodes, err := parseOutputNodes(&tpl.Outputs)
 	if err != nil {
-		return nil, fmt.Errorf("parse CloudFormation template %s: %w", template, err)
+		return nil, err
 	}
 
 	var outputs []Output
-	for logicalID, output := range tpl.Outputs {
-		outputs = append(outputs, Output{
-			Name:            logicalID,
-			IsSecret:        isSecret(output, tpl.GetAllSecretsManagerSecretResources()),
-			IsManagedPolicy: isManagedPolicy(output, tpl.GetAllIAMManagedPolicyResources()),
-		})
+	for _, outputNode := range outputNodes {
+		output := Output{
+			Name:            outputNode.name(),
+			IsSecret:        false,
+			IsManagedPolicy: false,
+		}
+		ref, ok := outputNode.ref()
+		if ok {
+			output.IsSecret = typeFor[ref] == secretManagerSecretType
+			output.IsManagedPolicy = typeFor[ref] == iamManagedPolicyType
+		}
+		outputs = append(outputs, output)
 	}
 	return outputs, nil
 }
 
-func isSecret(output cloudformation.Output, secrets map[string]*secretsmanager.Secret) bool {
-	value, ok := output.Value.(string)
-	if !ok {
-		return false
+// parseTypeByLogicalID returns a map where the key is the resource's logical ID and the value is the CloudFormation Type
+// of the resource such as "AWS::IAM::Role".
+func parseTypeByLogicalID(resourcesNode *yaml.Node) (typeFor map[string]string, err error) {
+	if resourcesNode.Kind != yaml.MappingNode {
+		// "Resources" is a required field in CloudFormation, check if it's defined as a map.
+		return nil, errors.New(`"Resources" field in cloudformation template is not a map`)
+	}
+
+	// The content of a !!map, like the "Resources" field, always come in pairs.
+	// The first element represents the key, ex: {Value: "Cluster", Kind: ScalarNode, Tag: "!!str", Content: nil}
+	// The second element holds the values such as "Type" and "Properties", ex: {Value: "", Kind: MappingNode, Tag:"!!map", Content:[...]}
+	typeFor = make(map[string]string)
+	for i := 0; i < len(resourcesNode.Content); i += 2 {
```

Comment:
this logic looks very familiar as https://github.com/aws/copilot-cli/blob/1b33c83e488047d5a7a42918bef4775464807519/internal/pkg/addon/cloudformation.go#L221-L226 and https://github.com/aws/copilot-cli/blob/e9e3994b0ad535be0627f2ce22b3f065e13d4e7e/internal/pkg/aws/cloudformation/parse.go#L31-L37. Can we possibly modularize and reuse this?

Reply:
💭 I think it's mostly looking bulky because of the redundant comments. Maybe I can do this:

```go
package yamlutil

func KeyValueMap(mappingNode *yaml.Node) map[*yaml.Node]*yaml.Node
```
But I don't want util packages 😭 especially with only one function.

I can create this function at least in this pkg since the same text is repeated three times. Let me know if you have any other ideas.

Reply:
A util pkg with full of all sorts of garbage function is all we need :)